File: RobHouses.java
import java.util.*;

// Time Complexity : o(n) 
// Space Complexity :o(1)

class Rob Houses {
    public static int rob(int arr[]){
        if(arr.length==0) return 0;
        if(arr.length==1) return arr[0];
        if(arr.length==2) return Math.max(arr[0],arr[1]);
        
        for(int i=2;i<arr.length;i++){
            arr[i] = arr[i] + ((i-3>=0)? Math.max(arr[i-2],arr[i-3]):arr[i-2]);
        }
       return Math.max(arr[arr.length-1], arr[arr.length-2]);
    }
  
	public static void main (String[] args) {
	  int arr[] = {1,2,3,1};
		System.out.println(rob(arr));
		
	}
}

File: robHouses.py
# Time Complexity :O(n)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : yes
# Any problem you faced while coding this : Yes, had to lookup solution several times.
class Solution:
    def rob(self, nums: List[int]) -> int:
        house1=0
        house2=0
        for n in nums:
            temp=max(house1+n,house2)
            house1,house2=house2,temp
        return house2